### PR TITLE
feat: RPC-fallback data source for Pipes

### PR DESCRIPTION
**Pull request by abernatskiy**
## RPC-fallback data source for Pipes

Adds an **RPC-backed EVM source** and a generic **fallback supervisor** to Pipes, mirroring the feature landed in `subsquid/squid-sdk` (PR #510). A pipe can now be fed by an ordered list of Portal and/or RPC sources; the supervisor drives the most-preferred healthy one, fails over on error/lag/staleness, and reclaims a recovered higher-preference source — all behind the same `AsyncIterable<PortalBatch> + pipeTo` a single Portal stream exposes, so it drops into an existing pipe unchanged.

### What's here
- **`FallbackSource`** — the supervisor: cascade-on-failure, resume-from-committed, `ForkException` pass-through, lag/staleness failover, global-stall hold, eager vs `onFailureOnly` reclaim, health + capability probing, and the single **monotonic finalized watermark** for the pipe (clamps every batch so a source switch can never un-finalize committed data).
- **`EvmRpcSource`** — an RPC EVM source whose output matches the Portal source's: delegates fetch/finality/continuity/fork detection to `@subsquid/evm-rpc`, then reuses the Portal decoder + a ported client-side filter (where-clause matching, relation expansion, augment/project-back of where-only fields, and **empty-block dropping** to match Portal output).
- **`createEvmFallback`** — builds a `FallbackSource` over Portal + RPC EVM sources sharing one field selection.
- **Metrics** — fallback health/switch/freshness gauges registered on the Pipes metrics server.
- **Dep hygiene** — `@subsquid/evm-rpc` + `@subsquid/evm-normalization` are **optional peer dependencies**, loaded lazily only when an `rpc` source becomes active; a Portal-only consumer never installs or imports the RPC stack.

### Feature parity with squid-sdk (this rebase)
Rebased onto current `main` (design/sdk1 merged) and audited against squid-sdk `master` in two passes — a test-scenario diff and a code-level feature diff. The two SDKs share no code, only scenarios.

**Ported test scenarios that had no Pipes counterpart** (all covered behavior that already existed but was untested):
- Supervisor: `onFailureOnly` never switches up; a `ForkException` from a source reached *after* a switch propagates; a slow downstream consumer between yields doesn't mark a source stale; probe-less eager reclaim via head-poll liveness alone.
- `SourceHealth`: a liveness pass resets the consecutive-fail counter; a failed capability probe flips unhealthy; a liveness pass while unhealthy doesn't reset the cooldown.
- evm/rpc filter: match traces by type, stateDiffs by address/kind, topic0-only logs, empty-where-matches-all, two-requests-kept-once, tx→siblings, log→transactionTraces, trace→parents; `toRequiredData` relation derivation; the **#510 merged-aggregate logs↔receipts dedup** regression; new `project.test.ts` (augment/selectionGrew) and `shim.test.ts`.
- Extracted the inline `withTimeout` + `safeReturn` into `core/fallback-async.ts` (mirrors squid's `fallback/async.ts`) with its edge-case unit tests.

**Behavioral fixes found during the audit:**
- **Empty-block parity** — the RPC source now drops interior blocks left empty by filtering (keeping batch boundaries / the progress cursor), matching Portal output. Previously it emitted every fetched block.
- **Missing-peer diagnostics** — the lazy RPC loader now discriminates a missing optional peer (ESM `ERR_MODULE_NOT_FOUND` / CJS `MODULE_NOT_FOUND` naming a peer) from any other load failure, which now surfaces unchanged instead of being masked as "peers missing".
- Added a supervisor-level guard that the finalized watermark clamp survives a source switch.

### Deliberately not ported (documented decisions)
- **Per-network presets** (squid ships 7-chain presets + a resolver + an `evmRpcStream` builder). This is a **public-API / product decision** for Pipes — it needs a URL→`Rpc` builder surface (squid's builder constructs the `Rpc` incl. validation config from a preset, whereas Pipes has the user construct `Rpc` and pass it). Left as a focused follow-up rather than guessing the API surface here. The underlying knobs (`RpcMethodOptions`, evm-rpc validation) are already exposed.
- **Capability-probe anchoring** (squid probes at `committed + lookahead` clamped to `head − tipMargin`). Pipes anchors the probe at the committed **finalized** cursor, which is already below the reorg-prone tip and in the same pruning regime it's about to read — achieving both goals squid's `lookahead`/`tipMargin` target, more simply, and without de-hashing the probe read. Kept as-is intentionally.

### Not applicable to Pipes (by design)
`getHead` delegation suite (Pipes' `FallbackSource` exposes no public `getHead`), `flattenRequest`/`unionRequiredData`/`streamBoundedRanges` (Pipes merges requests upstream and delegates streaming to `@subsquid/evm-rpc`), squid's megapackage packaging guard, and the LFS golden decode (covered by the gated parity e2e).

### Tests
501 non-service unit tests pass; `tsc --noEmit` clean. Live RPC/Portal parity remains behind the `RPC_E2E` gate.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/subsquid-pipes/package.json [hunk 1]
```diff
@@ -53,6 +53,8 @@
     "@google-cloud/bigquery": "8.3.0",
     "@google-cloud/bigquery-storage": "5.1.0",
     "@subsquid/borsh": "0.3.0",
+    "@subsquid/evm-normalization": "^0.0.2",
+    "@subsquid/evm-rpc": "^0.0.2",
     "@types/better-sqlite3": "7.6.13",
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/package.json:60]: **(With Claude):** Good catch — confirmed and fixed in 37df70c. `evm-rpc@0.0.2`'s peers (`http-client ^1.8.1`, `rpc-client ^4.15.1`) were unsatisfied because looser transitive ranges had already pinned 1.8.0 / 4.14.0. Pinned `^1.8.1` / `^4.15.1` in devDependencies (both current npm latests, and they satisfy every other consumer's range), so the lockfile now resolves `evm-rpc@0.0.2(@subsquid/http-client@1.8.1)(@subsquid/rpc-client@4.15.1)` with peers satisfied. 501 tests still green.

File: packages/subsquid-pipes/src/evm/rpc/filter.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    const filter = new FilterBuilder<AnyTrace>()
+    filter.propIn('type' as any, type)
+    filter.getIn((t) => (t.type === 'create' ? assertNotNull((t as any).action?.from) : undefined), createFrom)
+    filter.getIn((t) => (t.type === 'call' ? assertNotNull((t as any).action?.to) : undefined), callTo)
+    filter.getIn((t) => (t.type === 'call' ? assertNotNull((t as any).action?.from) : undefined), callFrom)
+    filter.getIn((t) => (t.type === 'call' ? assertNotNull((t as any).action?.sighash) : undefined), callSighash)
+    filter.getIn(
+      (t) => (t.type === 'suicide' ? assertNotNull((t as any).action?.refundAddress) : undefined),
+      suicideRefundAddress,
+    )
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/rpc/filter.ts]: **(With Claude):** Confirmed real (repro'd it — a sighash-less call trace threw), fixed in 48458fa. `FilterBuilder.getIn` already treats `undefined` as a non-match, so the `assertNotNull` bought nothing but the crash; all trace action fields are now read nullable-safely. Added regression tests for a sighash-less call and a `null` refundAddress. Note: the same pattern exists in squid-sdk's evm-rpc-stream filter — flagged for a matching fix there when that (currently paused) work resumes.

File: packages/subsquid-pipes/src/evm/rpc/decode.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * Force the structurally-required fields the cursor + filter engine always need (and which the
+ * Portal query always includes), so they survive `project` regardless of the user's selection.
+ */
+export function withRequiredFields(fields: FieldSelection): FieldSelection {
+  return {
+    block: { ...fields.block, number: true, hash: true, parentHash: true },
+    transaction: { ...fields.transaction, transactionIndex: true },
+    log: { ...fields.log, logIndex: true, transactionIndex: true },
+    trace: { ...fields.trace, transactionIndex: true, traceAddress: true },
+    stateDiff: { ...fields.stateDiff, transactionIndex: true, address: true, key: true },
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/rpc/decode.ts]: **(With Claude):** Fixed in fd94384. Added `kind: true` to the forced stateDiff selection in `withRequiredFields` (alongside address/key) so the tagged-union discriminator + kind where-clauses always survive projection. Unit-tested.

File: packages/subsquid-pipes/src/evm/evm-fallback.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+/**
+ * Build a {@link FallbackSource} over an ordered list of EVM sources (Portal and/or RPC), all
+ * sharing one field selection + request so they produce identical output. Drop-in for a single
+ * Portal stream — same `AsyncIterable<PortalBatch> + pipeTo`.
+ */
+export function createEvmFallback<F extends FieldSelection>(
+  options: EvmFallbackOptions<F>,
+): FallbackSource<Block<F>[]> {
+  const underlying: FallbackUnderlyingSource<Block<F>[]>[] = options.sources.map((cfg, i) => {
+    const source =
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/evm-fallback.ts:151]: **(With Claude):** Good catch — the feature really was unreachable. Fixed in fd94384: re-exported evm-fallback (`createEvmFallback`, `evmPortalReadSource`) from the `./evm` barrel. Kept it peer-free so a Portal-only `import '@subsquid/pipes/evm'` doesn't eagerly pull the optional evm-rpc peers — made decode.ts's `evm-normalization` import type-only (it was value syntax but used only as a type) and re-exported `RpcMethodOptions` type-only. The RPC stack is still lazy-loaded on first `rpc`-source read. Added a barrel-reachability test.

File: packages/subsquid-pipes/src/evm/rpc/project.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+  return false
+}
+
+/**
+ * Keep the items of `projected` whose positionally-aligned item in `pre` (the pre-filter decode of
+ * the same block) survived filtering into `kept`. `projected[i]` and `pre[i]` are the same on-chain
+ * item decoded at two field selections, so a surviving `pre[i]` means keep `projected[i]`. Position
+ * + identity — not a synthesized structural key — so items that would share such a key (block-reward
+ * traces carry no `transactionIndex`) can't collide. Mirrors the Squid evm-rpc-stream projection.
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/rpc/project.ts]: **(With Claude):** You're right — the schema marks `TraceBaseFields.transactionIndex` required, so the specific claim was wrong. Rephrased (here, in evm-rpc-source.ts, and the test) to the general point: position + object identity avoids confusing structurally identical items via a synthesized key. Fixed in fd94384's follow-up.

File: packages/subsquid-pipes/src/core/fallback-metrics.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    },
+  })
+
+  metrics.gauge({
+    name: `${prefix}_switches_total`,
+    help: 'Cumulative number of fallback source switches',
+    collect() {
+      this.set(source.metrics().switchCount)
+    },
+  })
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/core/fallback-metrics.ts:71]: **(With Claude):** Agreed — renamed to `sqd_fallback_switches` (dropped the `_total` suffix reserved for Counters; it's a pull-based gauge set to the current cumulative value). This also matches the Squid metric name. Test updated.

File: packages/subsquid-pipes/src/core/metrics-server.ts [hunk 7]
```diff
@@ -24,6 +24,8 @@ export interface CounterConfiguration<T extends string> extends MetricConfigurat
 export interface Gauge<T extends string = string> {
   set(value: number): void
   set(labels: LabelValues<T>, value: number): void
+  /** Clear all label combinations — call before re-setting in `collect` to drop stale series. */
+  reset(): void
 }
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/core/metrics-server.ts:33]: **(With Claude):** Good backward-compat catch — fixed in e07084a. Made `reset` optional on the public `Gauge` interface and optional-chained the sole caller (the fallback cause-label gauge). A custom `MetricsServer` that predates `reset()` now still satisfies the interface; without it, stale label series just aren't pruned (graceful degradation) rather than a type error.

File: packages/subsquid-pipes/src/evm/rpc/shim.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+export function shimWireBlock(block: any): any {
+  const traces = block?.traces
+  if (Array.isArray(traces)) {
+    for (const trace of traces) {
+      if (trace.type === 'selfdestruct') {
+        trace.type = 'suicide'
+      } else if (trace.type === 'reward' && trace.action && 'rewardType' in trace.action) {
+        trace.action.type = trace.action.rewardType
+        delete trace.action.rewardType
+      }
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/rpc/shim.ts:24]: **(With Claude):** Fixed in the latest push. Guarded the reward branch with a `typeof action === 'object' && action !== null` check before the `in`, so a malformed non-object `action` no longer throws — bad shapes fall through to `cast()` for a proper validation error. Added a test with a string `action`.

File: packages/subsquid-pipes/package.json [hunk 9]
```diff
@@ -82,6 +88,8 @@
     "@google-cloud/bigquery": "8.3.0",
     "@google-cloud/bigquery-storage": "5.1.0",
     "@opentelemetry/api": "^1.9.0",
+    "@subsquid/evm-normalization": "^0.0.2",
+    "@subsquid/evm-rpc": "^0.0.2",
     "better-sqlite3": "12.4.5",
     "drizzle-orm": "0.44.7",
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/package.json:96]: **(With Claude):** Fixed in 7926df4. Declared `@subsquid/http-client` (^1.8.1) and `@subsquid/rpc-client` (^4.15.1) — evm-rpc's direct peers — as optional peerDependencies, so installing the RPC stack is complete and pnpm stops warning. (evm-normalization has no direct peers; `@subsquid/logger` is only an *optional* peer of util-internal-validation, which pipes core already depends on, so it's pre-existing/out of scope.)

File: packages/subsquid-pipes/src/evm/evm-fallback.ts [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+  const err = e as NodeJS.ErrnoException
+  const isModuleNotFound = err?.code === 'ERR_MODULE_NOT_FOUND' || err?.code === 'MODULE_NOT_FOUND'
+  if (isModuleNotFound && RPC_PEERS.some((p) => err.message?.includes(`'${p}'`))) {
+    return new Error(
+      `RPC fallback source "${sourceName}" requires the optional peer dependencies ` +
+        `"${RPC_PEERS[0]}" and "${RPC_PEERS[1]}" — install them to use RPC sources, or use only 'portal' sources.`,
+    )
+  }
+  return err
+}
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/evm-fallback.ts]: **(With Claude):** Fixed in 7926df4. Added `@subsquid/http-client` and `@subsquid/rpc-client` to `translateMissingRpcPeer`'s detection list, and the message is now generated from the full peer list so it names all four packages. A consumer who installed only evm-rpc + evm-normalization and hits a transitive MODULE_NOT_FOUND now gets complete guidance. Added a regression test for the http-client case.

File: packages/subsquid-pipes/src/evm/browser.ts [hunk 11]
```diff
@@ -1,5 +1,9 @@
 export * from './abi/common.js'
 export * from './evm-decoder.js'
+// The RPC-fallback EVM source. Safe to export here: evm-fallback and its import chain reference the
+// optional evm-rpc / evm-normalization peers only as erased types — the RPC stack is loaded lazily
+// (dynamic import) the first time an `rpc` source is read, so a Portal-only consumer never pulls it.
+export * from './evm-fallback.js'
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/browser.ts]: **(With Claude):** Reworded in 75c04a8 — you're right the comment overclaimed. It now states the laziness is *runtime*-only: the RPC-config types (`EvmFallbackSourceConfig.rpc: Rpc`) are preserved in the .d.ts, so a TS consumer referencing them needs the peer installed for typechecking; Portal-only usage that never touches them doesn't. Full *type* decoupling would mean a separate subpath (e.g. `@subsquid/pipes/evm-rpc`) — that's a public-API-shape call, so I've left it as a noted follow-up rather than change the entrypoint here. Happy to do it if you'd prefer that surface.

File: packages/subsquid-pipes/src/evm/evm-fallback.ts [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+// Re-export the RPC method-selection type so consumers can type an `rpc` source's `method` without
+// importing from the lazily-loaded (optional-peer) evm-rpc-source module. Type-only: erased at build,
+// so it never eagerly pulls the optional peers.
+export type { RpcMethodOptions } from './evm-rpc-source.js'
+
+/** One EVM source in a fallback. Both kinds share the same `fields` + `request`. */
+export type EvmFallbackSourceConfig<F extends FieldSelection> =
+  | { type: 'portal'; name?: string; portal: string | PortalClientOptions | PortalClient }
+  | { type: 'rpc'; name?: string; rpc: Rpc; method?: RpcMethodOptions; strideSize?: number; strideConcurrency?: number }
+
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply packages/subsquid-pipes/src/evm/evm-fallback.ts:38]: **(With Claude):** Same fix (75c04a8) — the `export type` comment now says it's erased at *runtime* but preserved in the .d.ts, so TS users of the RPC-config types need @subsquid/evm-rpc. Runtime laziness (a Portal-only consumer never loading the peer) is intact. The only way to make the types *also* optional for `./evm` importers is to move the RPC-fallback surface behind its own subpath; flagged as a follow-up API decision rather than guessing it here (note it also slightly walks back the earlier 'export it from ./evm' fix, so worth a deliberate call).

**Comment by abernatskiy** [comment]
**(With Claude):** Superseded by #156 — the same fallback engine (health model, probes, staleness/lag detection, diagnostics, metrics), rebased across the package rename and restructured onto the single `evmStream` facade the SDK design asked for: sources plug in under `PortalStream` as block-stream clients, and the switching policy is now overridable as code (`fallback.strategy`) with the stock algorithm as the default. Closing in favor of #156.
